#About Dataset
The dataset contains 1338 observations (rows) and 7 features (columns). The dataset contains 4 numerical features (age, bmi, children and id) and 3 nominal features (sex, smoker and region) that were converted into factors with numerical value designated for each level.
But here, 'id' is irrelevant variable.

The purposes of this exercise to look into different features to observe their relationship, and plot a multiple linear regression based on several features of individual such as age, physical/family condition and location against their existing medical expense to be used for predicting future medical expenses of individuals that help medical insurance to make decision on charging the premium.

In [76]:
import warnings
warnings.filterwarnings('ignore')

In [77]:
import numpy as np
import pandas as pd
df=pd.read_csv("https://github.com/YBI-Foundation/Dataset/raw/main/Insurance%20Premium.csv")
df

ID  Age  Gender     BMI  Children Smoker Region  Premium
0        1   19  female  27.900         0    yes  south    16885
1        2   18    male  33.770         1     no   east     1726
2        3   28    male  33.000         3     no   east     4449
3        4   33    male  22.705         0     no   west    21984
4        5   32    male  28.880         0     no   west     3867
...    ...  ...     ...     ...       ...    ...    ...      ...
1333  1334   50    male  30.970         3     no   west    10601
1334  1335   18  female  31.920         0     no  north     2206
1335  1336   18  female  36.850         0     no   east     1630
1336  1337   21  female  25.800         0     no  south     2008
1337  1338   61  female  29.070         0    yes   west    29141

[1338 rows x 8 columns]

In [78]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   ID        1338 non-null   int64  
 1   Age       1338 non-null   int64  
 2   Gender    1338 non-null   object 
 3   BMI       1338 non-null   float64
 4   Children  1338 non-null   int64  
 5   Smoker    1338 non-null   object 
 6   Region    1338 non-null   object 
 7   Premium   1338 non-null   int64  
dtypes: float64(1), int64(4), object(3)
memory usage: 83.8+ KB


In [79]:
df.describe()

ID          Age          BMI     Children       Premium
count  1338.000000  1338.000000  1338.000000  1338.000000   1338.000000
mean    669.500000    39.207025    30.663397     1.094918  13270.414798
std     386.391641    14.049960     6.098187     1.205493  12110.012882
min       1.000000    18.000000    15.960000     0.000000   1122.000000
25%     335.250000    27.000000    26.296250     0.000000   4740.000000
50%     669.500000    39.000000    30.400000     1.000000   9382.000000
75%    1003.750000    51.000000    34.693750     2.000000  16640.000000
max    1338.000000    64.000000    53.130000     5.000000  63770.000000

In [80]:
df.isna().sum()

ID          0
Age         0
Gender      0
BMI         0
Children    0
Smoker      0
Region      0
Premium     0
dtype: int64

In [81]:
df.columns

Index(['ID', 'Age', 'Gender', 'BMI', 'Children', 'Smoker', 'Region',
       'Premium'],
      dtype='object')

In [82]:
y=df['Premium']
x=df[['Age', 'Gender', 'BMI', 'Children', 'Smoker', 'Region']]

In [83]:
x=pd.get_dummies(x,columns=['Gender','Region'],dtype=np.int32)
x

Age     BMI  Children Smoker  Gender_female  Gender_male  Region_east  \
0      19  27.900         0    yes              1            0            0   
1      18  33.770         1     no              0            1            1   
2      28  33.000         3     no              0            1            1   
3      33  22.705         0     no              0            1            0   
4      32  28.880         0     no              0            1            0   
...   ...     ...       ...    ...            ...          ...          ...   
1333   50  30.970         3     no              0            1            0   
1334   18  31.920         0     no              1            0            0   
1335   18  36.850         0     no              1            0            1   
1336   21  25.800         0     no              1            0            0   
1337   61  29.070         0    yes              1            0            0   

      Region_north  Region_south  Region_west  
0                0             1            0  
1                0             0            0  
2                0             0            0  
3                0             0            1  
4                0             0            1  
...            ...           ...          ...  
1333             0             0            1  
1334             1             0            0  
1335             0             0            0  
1336             0             1            0  
1337             0             0            1  

[1338 rows x 10 columns]

In [84]:
x['Smoker']=x['Smoker'].replace({'yes':0,'no':1})
x

Age     BMI  Children  Smoker  Gender_female  Gender_male  Region_east  \
0      19  27.900         0       0              1            0            0   
1      18  33.770         1       1              0            1            1   
2      28  33.000         3       1              0            1            1   
3      33  22.705         0       1              0            1            0   
4      32  28.880         0       1              0            1            0   
...   ...     ...       ...     ...            ...          ...          ...   
1333   50  30.970         3       1              0            1            0   
1334   18  31.920         0       1              1            0            0   
1335   18  36.850         0       1              1            0            1   
1336   21  25.800         0       1              1            0            0   
1337   61  29.070         0       0              1            0            0   

      Region_north  Region_south  Region_west  
0                0             1            0  
1                0             0            0  
2                0             0            0  
3                0             0            1  
4                0             0            1  
...            ...           ...          ...  
1333             0             0            1  
1334             1             0            0  
1335             0             0            0  
1336             0             1            0  
1337             0             0            1  

[1338 rows x 10 columns]

In [85]:
import statsmodels.api as sm
model=sm.OLS(y,x)
print(model.fit().summary())

                            OLS Regression Results                            
Dep. Variable:                Premium   R-squared:                       0.751
Model:                            OLS   Adj. R-squared:                  0.749
Method:                 Least Squares   F-statistic:                     500.8
Date:                Tue, 02 Jul 2024   Prob (F-statistic):               0.00
Time:                        13:46:08   Log-Likelihood:                -13548.
No. Observations:                1338   AIC:                         2.711e+04
Df Residuals:                    1329   BIC:                         2.716e+04
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Age             256.8569     11.899     21.587

###According to above EDA model, there is no irrelevant feature variable in the model.

In [86]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,random_state=2529)

In [87]:
from sklearn.linear_model import LinearRegression
lr=LinearRegression()
lr.fit(x_train,y_train)

LinearRegression()

In [88]:
y_pred=lr.predict(x_test)

In [89]:
from sklearn.metrics import r2_score, mean_absolute_error, mean_absolute_percentage_error
r2_score(y_test,y_pred)

0.7413246304402386

In [90]:
mean_absolute_error(y_test,y_pred)

4117.54817349984

In [91]:
mean_absolute_percentage_error(y_test,y_pred)

0.4382335884804565

###Since, r2 score = 74% means, all the feature variables are only able to explain 74% of the target variable of the model.
###Hence, there are some more feature variables required for better predictive model.